File: go-toys/config/testdata/cfgdir/.meta/validation/ConfigDirChecksum.py
import os
import hashlib
import array

start_dir = "./../.."
checksum = bytearray(32)

def process_dir(dir):
    for item in os.listdir(dir):
        if item.startswith("."): continue

        full_path = os.path.join(dir, item)

        if os.path.isdir(full_path):
            process_dir(full_path)
            continue

        relative_to_start = os.path.relpath(full_path, start_dir)
        hash = hashlib.sha256()
        hash.update(relative_to_start.encode("utf-8"))
        hash.update(bytearray(open(full_path, "rb").read()))
        item_hash = hash.digest()
        print( item_hash.hex() + " " + relative_to_start)
        xor_item( item_hash )

def xor_item(item_hash):
    global checksum
    checksum = bytes(a ^ b for (a, b) in zip(checksum, item_hash))

process_dir(start_dir)
print( "Checksum : " + checksum.hex() )

File: go-toys/config/testdata/cfgdir/.meta/validation/ConfigDirChecksum.js
const fs = require('fs'),
    path = require('path'),
    crypto = require('crypto');

let startDir = './../..'
let checksum = new Uint8Array(32)

let processDir = function (dir) {
    for (let item of fs.readdirSync(dir)) {
        if (item.startsWith('.')) {
            continue;
        }
        // get the item path
        let itemPath = path.join(dir, item);
        let stats = fs.statSync(itemPath);
        if (stats.isDirectory()) {
            processDir(itemPath);
            continue;
        }
        
        const hash = crypto.createHash('sha256');
        const relativeToStart = path.relative(startDir, itemPath)
        hash.update((new TextEncoder('utf-8')).encode(relativeToStart));
        hash.update(fs.readFileSync(itemPath));
        const itemHash = hash.digest();

        console.log(Buffer.from(itemHash).toString('hex'), relativeToStart );

        xorItemHash(itemHash)
    }
}

let xorItemHash = function (itemHash) {
    for (let i = 0; i < 32; i++) {
        checksum[i] = checksum[i] ^ itemHash[i]
    }
}

processDir(startDir);
console.log("Checksum :", Buffer.from(checksum).toString('hex'));
